In [0]:
Работу выполнили Мария Рахманова, София Плаксина, Анна Николаева,Алина Сульженко
import pandas as pd    # для работы с таблицами 
import numpy as np     # для работы с матрицами


import matplotlib      #для работы с картинками

import matplotlib.pyplot as plt
import seaborn as sns
plt.style.use('ggplot')  #стиль у графиков

%matplotlib inline

In [0]:
#подгружаем movies
movies=pd.read_csv('movies.csv', sep=',')
movies

movieId                                              title  \
0           1                                   Toy Story (1995)   
1           2                                     Jumanji (1995)   
2           3                            Grumpier Old Men (1995)   
3           4                           Waiting to Exhale (1995)   
4           5                 Father of the Bride Part II (1995)   
5           6                                        Heat (1995)   
6           7                                     Sabrina (1995)   
7           8                                Tom and Huck (1995)   
8           9                                Sudden Death (1995)   
9          10                                   GoldenEye (1995)   
10         11                     American President, The (1995)   
11         12                 Dracula: Dead and Loving It (1995)   
12         13                                       Balto (1995)   
13         14                                       Nixon (1995)   
14         15                            Cutthroat Island (1995)   
15         16                                      Casino (1995)   
16         17                       Sense and Sensibility (1995)   
17         18                                  Four Rooms (1995)   
18         19              Ace Ventura: When Nature Calls (1995)   
19         20                                 Money Train (1995)   
20         21                                  Get Shorty (1995)   
21         22                                     Copycat (1995)   
22         23                                   Assassins (1995)   
23         24                                      Powder (1995)   
24         25                           Leaving Las Vegas (1995)   
25         26                                     Othello (1995)   
26         27                                Now and Then (1995)   
27         28                                  Persuasion (1995)   
28         29  City of Lost Children, The (Cité des enfants p...   
29         30  Shanghai Triad (Yao a yao yao dao waipo qiao) ...   
...       ...                                                ...   
9712   188189                         Sorry to Bother You (2018)   
9713   188301                        Ant-Man and the Wasp (2018)   
9714   188675                                      Dogman (2018)   
9715   188751                Mamma Mia: Here We Go Again! (2018)   
9716   188797                                         Tag (2018)   
9717   188833              The Man Who Killed Don Quixote (2018)   
9718   189043                                  Boundaries (2018)   
9719   189111                                      Spiral (2018)   
9720   189333               Mission: Impossible - Fallout (2018)   
9721   189381                                    SuperFly (2018)   
9722   189547                                Iron Soldier (2010)   
9723   189713                              BlacKkKlansman (2018)   
9724   190183                           The Darkest Minds (2018)   
9725   190207                                        Tilt (2011)   
9726   190209                 Jeff Ross Roasts the Border (2017)   
9727   190213                                   John From (2015)   
9728   190215                                Liquid Truth (2017)   
9729   190219                                       Bunny (1998)   
9730   190221  Hommage à Zgougou (et salut à Sabine Mamou) (2...   
9731   191005                                     Gintama (2017)   
9732   193565                          Gintama: The Movie (2010)   
9733   193567  anohana: The Flower We Saw That Day - The Movi...   
9734   193571                                Silver Spoon (2014)   
9735   193573            Love Live! The School Idol Movie (2015)   
9736   193579           Jon Stewart Has Left the Building (2015)   
9737   193581          Black Butler: Book of the Atlantic (2017)   
9738   193583                       No Game No Life: Zero (201

In [0]:
#подгружаем ratings
ratings=pd.read_csv('ratings.csv', sep=',')
ratings

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
5            1       70     3.0   964982400
6            1      101     5.0   964980868
7            1      110     4.0   964982176
8            1      151     5.0   964984041
9            1      157     5.0   964984100
10           1      163     5.0   964983650
11           1      216     5.0   964981208
12           1      223     3.0   964980985
13           1      231     5.0   964981179
14           1      235     4.0   964980908
15           1      260     5.0   964981680
16           1      296     3.0   964982967
17           1      316     3.0   964982310
18           1      333     5.0   964981179
19           1      349     4.0   964982563
20           1      356     4.0   964980962
21           1      362     5.0   964982588
22           1      367     4.0   964981710
23           1      423     3.0   964982363
24           1      441     4.0   964980868
25           1      457     5.0   964981909
26           1      480     4.0   964982346
27           1      500     3.0   964981208
28           1      527     5.0   964984002
29           1      543     4.0   964981179
...        ...      ...     ...         ...
100806     610   150401     3.0  1479543210
100807     610   152077     4.0  1493845817
100808     610   152081     4.0  1493846503
100809     610   152372     3.5  1493848841
100810     610   155064     3.5  1493848456
100811     610   156371     5.0  1479542831
100812     610   156726     4.5  1493848444
100813     610   157296     4.0  1493846563
100814     610   158238     5.0  1479545219
100815     610   158721     3.5  1479542491
100816     610   158872     3.5  1493848024
100817     610   158956     3.0  1493848947
100818     610   159093     3.0  1493847704
100819     610   160080     3.0  1493848031
100820     610   160341     2.5  1479545749
100821     610   160527     4.5  1479544998
100822     610   160571     3.0  1493848537
100823     610   160836     3.0  1493844794
100824     610   161582     4.0  1493847759
100825     610   161634     4.0  1493848362
100826     610   162350     3.5  1493849971
100827     610   163937     3.5  1493848789
100828     610   163981     3.5  1493850155
100829     610   164179     5.0  1493845631
100830     610   166528     4.0  1493879365
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]

In [0]:
#датасеты
movies = movies.merge(ratings, on = 'movieId', how = 'inner')
movies

movieId                                              title  \
0             1                                   Toy Story (1995)   
1             1                                   Toy Story (1995)   
2             1                                   Toy Story (1995)   
3             1                                   Toy Story (1995)   
4             1                                   Toy Story (1995)   
5             1                                   Toy Story (1995)   
6             1                                   Toy Story (1995)   
7             1                                   Toy Story (1995)   
8             1                                   Toy Story (1995)   
9             1                                   Toy Story (1995)   
10            1                                   Toy Story (1995)   
11            1                                   Toy Story (1995)   
12            1                                   Toy Story (1995)   
13            1                                   Toy Story (1995)   
14            1                                   Toy Story (1995)   
15            1                                   Toy Story (1995)   
16            1                                   Toy Story (1995)   
17            1                                   Toy Story (1995)   
18            1                                   Toy Story (1995)   
19            1                                   Toy Story (1995)   
20            1                                   Toy Story (1995)   
21            1                                   Toy Story (1995)   
22            1                                   Toy Story (1995)   
23            1                                   Toy Story (1995)   
24            1                                   Toy Story (1995)   
25            1                                   Toy Story (1995)   
26            1                                   Toy Story (1995)   
27            1                                   Toy Story (1995)   
28            1                                   Toy Story (1995)   
29            1                                   Toy Story (1995)   
...         ...                                                ...   
100806   188301                        Ant-Man and the Wasp (2018)   
100807   188675                                      Dogman (2018)   
100808   188751                Mamma Mia: Here We Go Again! (2018)   
100809   188797                                         Tag (2018)   
100810   188833              The Man Who Killed Don Quixote (2018)   
100811   189043                                  Boundaries (2018)   
100812   189111                                      Spiral (2018)   
100813   189333               Mission: Impossible - Fallout (2018)   
100814   189333               Mission: Impossible - Fallout (2018)   
100815   189381                                    SuperFly (2018)   
100816   189547                                Iron Soldier (2010)   
100817   189713                              BlacKkKlansman (2018)   
100818   190183                           The Darkest Minds (2018)   
100819   190207                                        Tilt (2011)   
100820   190209                 Jeff Ross Roasts the Border (2017)   
100821   190213                                   John From (2015)   
100822   190215                                Liquid Truth (2017)   
100823   190219                                       Bunny (1998)   
100824   190221  Hommage à Zgougou (et salut à Sabine Mamou) (2...   
100825   191005                                     Gintama (2017)   
100826   193565                          Gintama: The Movie (2010)   
100827   193567  anohana: The Flower We Saw That Day - The Movi...   
100828   193571                                Silver Spoon (2014)   
100829   193573            Love Live! The School Idol Movie (2015)   
100830   193579           Jon Stewart Has Left the Building (2015)   
100831   193581   

In [0]:
#оставляем только нужные нам переменные
movies_SVD = movies[['userId','movieId','rating']]
movies_SVD

userId  movieId  rating
0            1        1     4.0
1            5        1     4.0
2            7        1     4.5
3           15        1     2.5
4           17        1     4.5
5           18        1     3.5
6           19        1     4.0
7           21        1     3.5
8           27        1     3.0
9           31        1     5.0
10          32        1     3.0
11          33        1     3.0
12          40        1     5.0
13          43        1     5.0
14          44        1     3.0
15          45        1     4.0
16          46        1     5.0
17          50        1     3.0
18          54        1     3.0
19          57        1     5.0
20          63        1     5.0
21          64        1     4.0
22          66        1     4.0
23          68        1     2.5
24          71        1     5.0
25          73        1     4.5
26          76        1     0.5
27          78        1     4.0
28          82        1     2.5
29          86        1     4.0
...        ...      ...     ...
100806     596   188301     4.0
100807     318   188675     3.5
100808     212   188751     4.5
100809     514   188797     4.0
100810     318   188833     4.5
100811     338   189043     2.5
100812     338   189111     3.0
100813     184   189333     4.0
100814     248   189333     3.5
100815     318   189381     2.5
100816     210   189547     1.0
100817     462   189713     2.5
100818      50   190183     3.5
100819     338   190207     1.5
100820     338   190209     4.0
100821     338   190213     1.0
100822     338   190215     1.5
100823     338   190219     1.0
100824     338   190221     1.0
100825     184   191005     4.5
100826     184   193565     3.5
100827     184   193567     3.0
100828     184   193571     4.0
100829     184   193573     4.0
100830     184   193579     3.5
100831     184   193581     4.0
100832     184   193583     3.5
100833     184   193585     3.5
100834     184   193587     3.5
100835     331   193609     4.0

[100836 rows x 3 columns]

In [0]:
from surprise import Dataset, Reader
reader=Reader(rating_scale=(1,5))
data = Dataset.load_from_df(movies_SVD, reader)

In [0]:
from surprise import SVD #подгружаем методом SVD матрицы
algo = SVD(n_factors=100, n_epochs=100, lr_all=0.005, reg_all=0.1)

In [0]:
from surprise.model_selection import cross_validate
cross_validate(algo, data, measures=['MSE', 'MAE'], cv=5, verbose=True)

Evaluating MSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
MSE (testset)     0.7280  0.7254  0.7222  0.7105  0.7261  0.7224  0.0062  
MAE (testset)     0.6546  0.6530  0.6513  0.6464  0.6499  0.6510  0.0028  
Fit time          28.36   28.18   28.03   28.64   28.41   28.33   0.21    
Test time         0.21    0.21    0.15    0.20    0.21    0.20    0.02    


{'test_mse': array([0.72797513, 0.7253823 , 0.72223136, 0.71052265, 0.72611209]),
 'test_mae': array([0.65456998, 0.6529796 , 0.65132953, 0.6463738 , 0.64993843]),
 'fit_time': (28.364932537078857,
  28.17983102798462,
  28.03192138671875,
  28.639491319656372,
  28.411866426467896),
 'test_time': (0.20844578742980957,
  0.20600223541259766,
  0.1495976448059082,
  0.19844412803649902,
  0.21494245529174805)}

Наше MSE=0.7224
MAE=0.6510
Наша команда использовала метод SVD матрицы. 

Источник: https://github.com/NicolasHug/Surprise
        